### PR TITLE
Fix performance regression for device compileAndRun

### LINKED ISSUE #59: Performance regression for overhead in `compileAndRun`
For convenience we extract the types of runtime arguments from the bitcode. This means every time we need to parse the IR and extract them. We should avoid this overhead

### PR DESCRIPTION
**Pull request by ggeorgakoudis**
- Extract runtime constant types in the pass
- Avoid bitcode parsing unless compiling
- Link binaries for CUDA RDC in object codegen
- Add kernel repeat test

Closes #59

### CODE REVIEW

File: lib/CompilerInterfaceTypes.h [hunk 1]
```diff
@@ -14,6 +14,19 @@
 #include <cstring>
 #include <stdint.h>
 
+namespace proteus {
+
+enum RuntimeConstantTypes : int32_t {
+  BOOL = 1,
```
Comment (johnbowen42) [inline lib/CompilerInterfaceTypes.h:20]: This is a great addition, I should've done this in the first place

File: lib/CompilerInterfaceTypes.h [hunk 2]
```diff
@@ -14,6 +14,19 @@
 #include <cstring>
 #include <stdint.h>
 
+namespace proteus {
+
+enum RuntimeConstantTypes : int32_t {
```
Comment (johnbowen42) [inline lib/CompilerInterfaceTypes.h:19]: why inherit from, int32_t, just to enforce size?

File: lib/JitEngineDevice.hpp [hunk 3]
```diff
@@ -51,20 +53,22 @@ using namespace llvm;
...
-      RCIndices.push_back(_RCIndices[I]);
-    }
-  }
-  JITKernelInfo() : Name(nullptr), NumRCs(0), RCIndices() {}
-  auto getName() const { return Name; }
-  auto getRCIndices() const { return RCIndices; }
-  auto getNumRCs() const { return NumRCs; }
+  JITKernelInfo(char const *Name, int32_t *RCIndices, int32_t *RCTypes,
+                int32_t NumRCs)
+      : Name(Name), RCIndices{ArrayRef{RCIndices, static_cast<size_t>(NumRCs)}},
```
Comment (johnbowen42) [inline lib/JitEngineDevice.hpp:63]: nice use of the constructors here

**Review by johnbowen42 — APPROVED** [review]
LGTM.  This improves not only performance but also readability and simplifies the code IMO.

File: lib/CompilerInterfaceTypes.h (hunk 1 above)
Comment (ggeorgakoudis) [reply lib/CompilerInterfaceTypes.h:20]: Yes, to match the expected data type from the pass

File: lib/CompilerInterfaceTypes.h (hunk 2 above)
Comment (ggeorgakoudis) [reply lib/CompilerInterfaceTypes.h:19]: Yes, to match the expected data type from the pass

**Review by tbennun — APPROVED** [review]
LGTM, solves my performance issue as well.

**Pull request merged**